### ISSUE TITLE
xmlrpc calls Request.finish a second time

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4423]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ivank reported|
|-|-|
|Trac ID|trac#4423|
|Type|defect|
|Created|2010-05-07 00:41:31Z|
|Branch|https://github.com/twisted/twisted/tree/xmlrpc-finish-after-error-4423|

`twisted.web.xmlrpc` needs to use `Request.notifyFinish` to track if the Request is already finished, and if so, not write to or finish the request.

psykidellic linked to http://bpaste.net/show/6093/ which shows the typical "RuntimeError: Request.finish called on a request after its connection was lost; use Request.notifyFinish to keep track of this."

<details><summary>Searchable metadata</summary>

```
trac-id__4423 4423
type__defect defect
reporter__ivank ivank
priority__high high
milestone__ 
branch__branches_xmlrpc_finish_after_error_4423 branches/xmlrpc-finish-after-error-4423
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__easy easy
time__1273192891000000 1273192891000000
changetime__1274554848000000 1274554848000000
version__None None
owner__ 
cc__psykidellic
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [28949]) Branching to 'xmlrpc-finish-after-error-4423'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [28950]) Use notifyFinish to avoid writing the response when no one is listening

refs [#4423](https://github.com/twisted/twisted/issues/4423)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @exarkun|
|-|-|

Looks pretty good, except ... 3 space indents!??  What the heck!

Please fix the indentation, then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [28983]) Merge xmlrpc-finish-after-error-4423

Author: exarkun
Reviewer: glyph
Fixes: [#4423](https://github.com/twisted/twisted/issues/4423)

Keep track of whether the request has failed when generating an XML-RPC response.
If it does fail before the response is ready to be sent, don't bother trying
to send it.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
